### PR TITLE
recreate_index breakdown in celery subtasks

### PR DESCRIPTION
**Pull request by arslanashraf7**
#### Pre-Flight checklist

⚠️  **NOTE:** This PR is rebased on https://github.com/mitodl/micromasters/pull/4960

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
https://github.com/mitodl/micromasters/issues/4959

#### What's this PR do?
✔️  **NOTE:** The implementation is inspired by [reindexing](https://github.com/mitodl/open-discussions/blob/master/search/tasks.py#L699) in Open-Discussions

- Converts the `recreate_index` into a celery task
- Divides the indexing operation into possible celery subtasks to run parallel
- Introduces `ELASTICSEARCH_INDEXING_CHUNK_SIZE` and runs multiple celery tasks of  Program User Enrollment and Percolate Queries
- Each subtask indices `ELASTICSEARCH_INDEXING_CHUNK_SIZE` this many elements

#### How should this be manually tested?
- Run the `recreate_index` command, this should now start a celery task that further starts 2 types of celery subtasks for indexing enrollments and percolates. 
- Set a test value for `ELASTICSEARCH_INDEXING_CHUNK_SIZE` (Default is 100)
- On running the command you should see multiple celery tasks sent to perform indexing on chunks of data(chunk size would be `ELASTICSEARCH_INDEXING_CHUNK_SIZE`)
- Make sure that the number of celery tasks matches the chunk size division above
- Check the parent task logs the task details which are traceable. e.g if the task fails we can investigate

#### Where should the reviewer start?
- `recreate_index` command

#### Any background context you want to provide?
- We used to run reindexing as a synchronous process before, and we were facing timeouts during the process when the recreate_index was run on a lot of data
- We did some related updates in [Open Discussions](https://github.com/mitodl/open-discussions/blob/master/search/tasks.py#L699) and it made sense to update micromasters also to make reindexing faster

### CODE REVIEW

**Comment by arslanashraf7** [comment]
@rhysyngsun, This is the second PR as part of what was discussed [here](https://github.com/mitodl/micromasters/issues/4959#issuecomment-866795414). The first PR of the reindexing process conversion to celery was [this](https://github.com/mitodl/micromasters/pull/4960).

Please share your thoughts about what you think of the feasibility of this PR.

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
The code generally looks good and the reindex succeeded for me, but my local setup appears to be broken when I try to access the site and I've had difficulty fixing it, could you tap someone else to make sure the `/learners` ui still works as expected?

File: micromasters/utils.py [hunk 1]
```diff
@@ -281,3 +281,34 @@ def generate_md5(unicode):
     hasher = hashlib.md5()
     hasher.update(unicode)
     return hasher.hexdigest()
+
+
+def merge_strings(list_or_str):
```
[the remark this answers is not in the document]
Comment (arslanashraf7) [reply micromasters/utils.py:286]: Added.

**Comment by arslanashraf7** [comment]
> The code generally looks good and the reindex succeeded for me, but my local setup appears to be broken when I try to access the site and I've had difficulty fixing it, could you tap someone else to make sure the `/learners` ui still works as expected?

@annagav or @noisecapella do you happen to have a locally running MicroMasters instance. If yes, would you be able to take it from here?

**Review by noisecapella — APPROVED** [review]

**Comment by noisecapella** [comment]
> > The code generally looks good and the reindex succeeded for me, but my local setup appears to be broken when I try to access the site and I've had difficulty fixing it, could you tap someone else to make sure the `/learners` ui still works as expected?
> 
> @annagav or @noisecapella do you happen to have a locally running MicroMasters instance. If yes, would you be able to take it from here?

Yes, I just deleted and ran the reindex and it works for me

**Comment by arslanashraf7** [comment]
> Yes, I just deleted and ran the reindex and it works for me

@noisecapella Thanks for the review, There is one post PR issue regarding Heroku which I'm mentioning below.

- The review app for this PR times out on post-deploy execution when [this script](https://github.com/mitodl/micromasters/blob/master/app.json#L264) runs. The Heroku logs for this PR can be seen [here](https://dashboard.heroku.com/apps/micromasters-ci-pr-4968). Additionally [this](https://github.com/mitodl/micromasters/pull/3640/files#diff-96677aa35debfefd031d9d34d9c70369754ee3acb2d9a9d4090e98612efee6f5R242) is the reference to when we started running the recreate index as part of post-deployment
- I also noticed that we aren't running `recreate_index` in [open-discussions](https://github.com/mitodl/open-discussions/blob/master/app.json#L623) that's why we never saw the timeout issue there.

Based on the above, what would your thoughts be? Should we stop running `recreate_index` as part of the post-deployment script? or do we need to do some tweaks to our dynos on Heroku?

**Comment by noisecapella** [comment]
@arslanashraf7 I think it makes sense to remove `./manage.py recreate_index` from `app.json` and just run it manually on the review apps where we need to

**Pull request merged**